File: backend/web/extract.go
package web

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	neturl "net/url"
	"regexp"
	"strings"
	"time"

	readability "codeberg.org/readeck/go-readability/v2"
	"golang.org/x/net/html"
)

var (
	metaDescRe = regexp.MustCompile(`(?i)<meta[^>]+name=["']description["'][^>]+content=["']([^"']+)["']`)
	ogDescRe   = regexp.MustCompile(`(?i)<meta[^>]+property=["']og:description["'][^>]+content=["']([^"']+)["']`)
)

type htmlFetch struct {
	finalURL string
	status   int
	body     []byte
}

func downloadHTML(ctx context.Context, pageURL string) (htmlFetch, *FetchError) {
	timeout := time.Duration(fetchTimeoutSec) * time.Second
	reqCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, pageURL, nil)
	if err != nil {
		return htmlFetch{}, &FetchError{Kind: FetchNetwork, Message: err.Error()}
	}
	req.Header.Set("User-Agent", userAgent())

	client := &http.Client{Timeout: timeout}
	resp, err := client.Do(req)
	if err != nil {
		if reqCtx.Err() != nil {
			return htmlFetch{}, &FetchError{Kind: FetchTimeout, Message: "request timed out"}
		}
		return htmlFetch{}, &FetchError{Kind: FetchNetwork, Message: err.Error()}
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(io.LimitReader(resp.Body, 4<<20))
	if err != nil {
		return htmlFetch{}, &FetchError{Kind: FetchNetwork, Message: err.Error()}
	}

	status := resp.StatusCode
	if status == http.StatusTooManyRequests {
		return htmlFetch{}, &FetchError{Kind: FetchRateLimited, HTTPStatus: status, Message: "rate limited"}
	}
	if status == http.StatusForbidden || status == http.StatusUnauthorized {
		return htmlFetch{}, &FetchError{Kind: FetchBlocked, HTTPStatus: status, Message: "access denied"}
	}
	if status >= 400 {
		return htmlFetch{}, &FetchError{Kind: FetchHTTPError, HTTPStatus: status, Message: http.StatusText(status)}
	}

	finalURL := pageURL
	if resp.Request != nil && resp.Request.URL != nil {
		finalURL = resp.Request.URL.String()
	}
	return htmlFetch{finalURL: finalURL, status: status, body: body}, nil
}

func ExtractPageContent(pageURL string, body []byte) (title, content string, err *FetchError) {
	if text, t, ok := extractYouTubeFromHTML(pageURL, body); ok {
		return t, text, nil
	}

	parsed, parseErr := neturl.Parse(pageURL)
	if parseErr != nil {
		parsed, _ = neturl.ParseRequestURI(pageURL)
	}

	article, readErr := readability.FromReader(bytes.NewReader(body), parsed)
	if readErr == nil && article.Node != nil {
		var buf bytes.Buffer
		if renderErr := article.RenderText(&buf); renderErr == nil {
			text := strings.TrimSpace(buf.String())
			if text != "" {
				return article.Title(), text, nil
			}
		}
	}

	if meta := extractMetaDescription(body); meta != "" {
		title := extractHTMLTitle(body)
		return title, meta, nil
	}

	if plain := stripHTMLText(body); len(plain) >= 120 {
		return extractHTMLTitle(body), plain, nil
	}

	if readErr == nil && article.Node == nil {
		return "", "", &FetchError{
			Kind:    FetchJSRendered,
			Message: "no article content in static HTML — page likely requires JavaScript",
		}
	}
	if readErr != nil && strings.Contains(readErr.Error(), "Node field is nil") {
		return "", "", &FetchError{
			Kind:    FetchJSRendered,
			Message: "no article content in static HTML — page likely requires JavaScript",
		}
	}

	return "", "", &FetchError{Kind: FetchNoContent, Message: "no readable content extracted"}
}

func extractMetaDescription(body []byte) string {
	s := string(body)
	if m := ogDescRe.FindStringSubmatch(s); len(m) == 2 {
		return strings.TrimSpace(htmlUnescape(m[1]))
	}
	if m := metaDescRe.FindStringSubmatch(s); len(m) == 2 {
		return strings.TrimSpace(htmlUnescape(m[1]))
	}
	return ""
}

func extractHTMLTitle(body []byte) string {
	doc, err := html.Parse(bytes.NewReader(body))
	if err != nil {
		return ""
	}
	var title string
	var walk func(*html.Node)
	walk = func(n *html.Node) {
		if title != "" {
			return
		}
		if n.Type == html.ElementNode && n.Data == "title" && n.FirstChild != nil {
			title = strings.TrimSpace(n.FirstChild.Data)
			return
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			walk(c)
		}
	}
	walk(doc)
	return title
}

func stripHTMLText(body []byte) string {
	doc, err := html.Parse(bytes.NewReader(body))
	if err != nil {
		return ""
	}
	removeNodes(doc, "script", "style", "noscript", "nav", "header", "footer")
	var buf strings.Builder
	collectText(doc, &buf)
	return strings.Join(strings.Fields(buf.String()), " ")
}

func removeNodes(n *html.Node, tags ...string) {
	if n == nil {
		return
	}
	tagSet := map[string]bool{}
	for _, t := range tags {
		tagSet[t] = true
	}
	var prune func(*html.Node)
	prune = func(node *html.Node) {
		for c := node.FirstChild; c != nil; {
			next := c.NextSibling
			if c.Type == html.ElementNode && tagSet[c.Data] {
				node.RemoveChild(c)
			} else {
				prune(c)
			}
			c = next
		}
	}
	prune(n)
}

func collectText(n *html.Node, buf *strings.Builder) {
	if n.Type == html.TextNode {
		t := strings.TrimSpace(n.Data)
		if t != "" {
			if buf.Len() > 0 {
				buf.WriteByte(' ')
			}
			buf.WriteString(t)
		}
	}
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		collectText(c, buf)
	}
}

func htmlUnescape(s string) string {
	return html.UnescapeString(s)
}

func extractYouTubeFromHTML(pageURL string, body []byte) (text, title string, ok bool) {
	u, err := neturl.Parse(pageURL)
	if err != nil || !isYouTubeHost(u.Hostname()) {
		return "", "", false
	}
	title = extractHTMLTitle(body)
	desc := extractMetaDescription(body)
	if title == "" && desc == "" {
		return "", "", false
	}
	var b strings.Builder
	if title != "" {
		b.WriteString("Title: ")
		b.WriteString(title)
		b.WriteByte('\n')
	}
	if desc != "" {
		b.WriteString("Description: ")
		b.WriteString(desc)
		b.WriteByte('\n')
	}
	b.WriteString("\nNote: Full YouTube transcripts and comments require the watch page API or a browser; static fetch returns metadata only.")
	return strings.TrimSpace(b.String()), title, true
}

func isYouTubeHost(host string) bool {
	host = strings.ToLower(host)
	return host == "youtube.com" || host == "www.youtube.com" || host == "m.youtube.com" || host == "youtu.be"
}

func fetchYouTubeOEmbed(ctx context.Context, pageURL string) (PageHit, error) {
	oembed := "https://www.youtube.com/oembed?format=json&url=" + neturl.QueryEscape(pageURL)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, oembed, nil)
	if err != nil {
		return PageHit{}, err
	}
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return PageHit{}, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return PageHit{}, fmt.Errorf("youtube oembed: %s", resp.Status)
	}
	var data struct {
		Title      string `json:"title"`
		AuthorName string `json:"author_name"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return PageHit{}, err
	}
	var b strings.Builder
	fmt.Fprintf(&b, "Title: %s\n", data.Title)
	if data.AuthorName != "" {
		fmt.Fprintf(&b, "Channel: %s\n", data.AuthorName)
	}
	b.WriteString("\nNote: YouTube descriptions and comments are not available via static fetch.")
	return PageHit{Title: data.Title, URL: pageURL, Content: b.String()}, nil
}

File: backend/web/extract.ts
// PORT: backend/web/extract.go

import { Effect } from "effect";
import {
  type page_hit,
  fetch_error,
  fetch_failure_kind,
  fetch_timeout_sec,
} from "./types";
import { user_agent } from "./url_guard";

const metaDescRe = /<meta[^>]+name=["']description["'][^>]+content=["']([^"']+)["']/i;
const ogDescRe = /<meta[^>]+property=["']og:description["'][^>]+content=["']([^"']+)["']/i;

export type html_fetch = {
  finalURL: string;
  status: number;
  body: string;
};

export const download_html = (
  ctx: AbortSignal,
  pageURL: string,
): Effect.Effect<html_fetch, fetch_error> => {
  return Effect.tryPromise({
    try: async () => {
      const controller = new AbortController();
      const onAbort = () => controller.abort();
      ctx.addEventListener("abort", onAbort);

      const timeoutId = setTimeout(() => controller.abort(), fetch_timeout_sec * 1000);

      try {
        const resp = await fetch(pageURL, {
          method: "GET",
          headers: {
            "User-Agent": user_agent(),
          },
          signal: controller.signal,
        });

        clearTimeout(timeoutId);
        ctx.removeEventListener("abort", onAbort);

        const status = resp.status;
        if (status === 429) {
          throw new fetch_error(fetch_failure_kind.fetch_rate_limited, status, "rate limited");
        }
        if (status === 403 || status === 401) {
          throw new fetch_error(fetch_failure_kind.fetch_blocked, status, "access denied");
        }
        if (status >= 400) {
          throw new fetch_error(fetch_failure_kind.fetch_http_error, status, `HTTP error ${status}`);
        }

        const bodyText = await resp.text();
        return {
          finalURL: resp.url || pageURL,
          status,
          body: bodyText,
        };
      } catch (err: any) {
        clearTimeout(timeoutId);
        ctx.removeEventListener("abort", onAbort);

        if (err instanceof fetch_error) {
          throw err;
        }

        if (controller.signal.aborted) {
          throw new fetch_error(fetch_failure_kind.fetch_timeout, 0, "request timed out");
        }
        throw new fetch_error(fetch_failure_kind.fetch_network, 0, err.message || String(err));
      }
    },
    catch: (cause) => {
      if (cause instanceof fetch_error) {
        return cause;
      }
      return new fetch_error(fetch_failure_kind.fetch_network, 0, String(cause));
    },
  });
};

type token =
  | { type: "tag-start"; name: string; isClose: boolean; isSelfClosing: boolean }
  | { type: "text"; text: string }
  | { type: "comment" };

const tokenize_html = (html: string): token[] => {
  const tokens: token[] = [];
  let i = 0;
  while (i < html.length) {
    if (html[i] === "<") {
      if (html.slice(i, i + 4) === "<!--") {
        const closeIdx = html.indexOf("-->", i + 4);
        if (closeIdx === -1) {
          i = html.length;
        } else {
          tokens.push({ type: "comment" });
          i = closeIdx + 3;
        }
        continue;
      }
      const closeIdx = html.indexOf(">", i);
      if (closeIdx === -1) {
        tokens.push({ type: "text", text: html.slice(i) });
        break;
      }
      const tagContent = html.slice(i + 1, closeIdx).trim();
      i = closeIdx + 1;
      if (tagContent === "") continue;

      const isClose = tagContent.startsWith("/");
      const isSelfClosing = tagContent.endsWith("/");
      let tagName = tagContent;
      if (isClose) tagName = tagName.slice(1);
      if (isSelfClosing) tagName = tagName.slice(0, -1);
      tagName = tagName.split(/\s+/)[0].toLowerCase();

      tokens.push({ type: "tag-start", name: tagName, isClose, isSelfClosing });
    } else {
      const nextOpen = html.indexOf("<", i);
      if (nextOpen === -1) {
        tokens.push({ type: "text", text: html.slice(i) });
        break;
      }
      tokens.push({ type: "text", text: html.slice(i, nextOpen) });
      i = nextOpen;
    }
  }
  return tokens;
};

type parse_node = {
  type: "element" | "text";
  name: string;
  text?: string;
  children: parse_node[];
};

const build_html_tree = (tokens: token[]): parse_node => {
  const root: parse_node = { type: "element", name: "root", children: [] };
  const stack: parse_node[] = [root];

  for (const tok of tokens) {
    if (tok.type === "comment") {
      continue;
    }
    if (tok.type === "text") {
      if (stack.length > 0) {
        stack[stack.length - 1].children.push({
          type: "text",
          name: "",
          text: tok.text,
          children: [],
        });
      }
    } else if (tok.type === "tag-start") {
      if (tok.isClose) {
        const idx = stack.map((n) => n.name).lastIndexOf(tok.name);
        if (idx > 0) {
          stack.splice(idx);
        }
      } else {
        const node: parse_node = {
          type: "element",
          name: tok.name,
          children: [],
        };
        if (stack.length > 0) {
          stack[stack.length - 1].children.push(node);
        }
        const selfClosingTags = new Set(["img", "br", "hr", "input", "meta", "link"]);
        if (!tok.isSelfClosing && !selfClosingTags.has(tok.name)) {
          stack.push(node);
        }
      }
    }
  }
  return root;
};

const remove_nodes = (node: parse_node, tags: Set<string>): void => {
  node.children = node.children.filter((child) => {
    if (child.type === "element" && tags.has(child.name)) {
      return false;
    }
    remove_nodes(child, tags);
    return true;
  });
};

const collect_text = (node: parse_node, parts: string[]): void => {
  if (node.type === "text" && node.text) {
    const trimmed = node.text.trim();
    if (trimmed !== "") {
      parts.push(trimmed);
    }
  }
  for (const child of node.children) {
    collect_text(child, parts);
  }
};

export const strip_html_text = (htmlStr: string): string => {
  const tokens = tokenize_html(htmlStr);
  const root = build_html_tree(tokens);
  remove_nodes(root, new Set(["script", "style", "noscript", "nav", "header", "footer"]));
  const parts: string[] = [];
  collect_text(root, parts);
  return parts.join(" ").replace(/\s+/g, " ").trim();
};

export const extract_html_title = (htmlStr: string): string => {
  const tokens = tokenize_html(htmlStr);
  const root = build_html_tree(tokens);

  let title = "";
  const findTitle = (node: parse_node) => {
    if (title !== "") return;
    if (node.type === "element" && node.name === "title") {
      const parts: string[] = [];
      collect_text(node, parts);
      title = parts.join(" ").trim();
      return;
    }
    for (const child of node.children) {
      findTitle(child);
    }
  };
  findTitle(root);
  return title;
};

export const html_unescape = (s: string): string => {
  return s
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&apos;/g, "'")
    .replace(/&#(\d+);/g, (_, dec) => String.fromCharCode(parseInt(dec, 10)))
    .replace(/&#x([0-9a-f]+);/gi, (_, hex) => String.fromCharCode(parseInt(hex, 16)));
};

export const extract_meta_description = (body: string): string => {
  let m = ogDescRe.exec(body);
  if (m && m.length === 2) {
    return html_unescape(m[1].trim());
  }
  m = metaDescRe.exec(body);
  if (m && m.length === 2) {
    return html_unescape(m[1].trim());
  }
  return "";
};

export const is_youtube_host = (host: string): boolean => {
  const h = host.toLowerCase();
  return (
    h === "youtube.com" ||
    h === "www.youtube.com" ||
    h === "m.youtube.com" ||
    h === "youtu.be"
  );
};

export const extract_youtube_from_html = (
  pageURL: string,
  body: string,
): [string, string, boolean] => {
  try {
    const u = new URL(pageURL);
    if (!is_youtube_host(u.hostname)) {
      return ["", "", false];
    }
    const title = extract_html_title(body);
    const desc = extract_meta_description(body);
    if (title === "" && desc === "") {
      return ["", "", false];
    }
    let out = "";
    if (title !== "") {
      out += `Title: ${title}\n`;
    }
    if (desc !== "") {
      out += `Description: ${desc}\n`;
    }
    out +=
      "\nNote: Full YouTube transcripts and comments require the watch page API or a browser; static fetch returns metadata only.";
    return [out.trim(), title, true];
  } catch {
    return ["", "", false];
  }
};

export const fetch_youtube_oembed = (
  ctx: AbortSignal,
  pageURL: string,
): Effect.Effect<page_hit, Error> => {
  return Effect.tryPromise({
    try: async () => {
      const oembed = `https://www.youtube.com/oembed?format=json&url=${encodeURIComponent(
        pageURL,
      )}`;
      const resp = await fetch(oembed, { signal: ctx });
      if (resp.status !== 200) {
        throw new Error(`youtube oembed: HTTP ${resp.status}`);
      }
      const data = (await resp.json()) as { title: string; author_name?: string };
      let content = `Title: ${data.title}\n`;
      if (data.author_name) {
        content += `Channel: ${data.author_name}\n`;
      }
      content += "\nNote: YouTube descriptions and comments are not available via static fetch.";
      return {
        title: data.title,
        url: pageURL,
        content,
        fetch: null,
      };
    },
    catch: (cause) => cause instanceof Error ? cause : new Error(String(cause)),
  });
};

const extract_readability_text = (root: parse_node): string => {
  remove_nodes(
    root,
    new Set(["script", "style", "noscript", "nav", "header", "footer", "aside", "iframe"]),
  );

  const paragraphs: string[] = [];
  const collectParagraphs = (node: parse_node) => {
    if (node.type === "element" && node.name === "p") {
      const parts: string[] = [];
      collect_text(node, parts);
      const txt = parts.join(" ").trim();
      if (txt !== "") {
        paragraphs.push(txt);
      }
      return;
    }
    for (const child of node.children) {
      collectParagraphs(child);
    }
  };
  collectParagraphs(root);

  if (paragraphs.length > 0) {
    const text = paragraphs.join("\n\n");
    if (text.length >= 200) {
      return text;
    }
  }

  const parts: string[] = [];
  collect_text(root, parts);
  return parts.join(" ").replace(/\s+/g, " ").trim();
};

export const extract_page_content = (
  pageURL: string,
  body: string,
): [string, string, fetch_error | null] => {
  const [ytText, ytTitle, ytOk] = extract_youtube_from_html(pageURL, body);
  if (ytOk) {
    return [ytTitle, ytText, null];
  }

  const tokens = tokenize_html(body);
  const root = build_html_tree(tokens);

  const title = extract_html_title(body);
  const content = extract_readability_text(root);

  if (content !== "") {
    return [title, content, null];
  }

  const meta = extract_meta_description(body);
  if (meta !== "") {
    return [title, meta, null];
  }

  const plain = strip_html_text(body);
  if (plain.length >= 120) {
    return [title, plain, null];
  }

  if (body.includes("<script") && plain.length < 120) {
    return [
      "",
      "",
      new fetch_error(
        fetch_failure_kind.fetch_js_rendered,
        0,
        "no article content in static HTML — page likely requires JavaScript",
      ),
    ];
  }

  return [
    "",
    "",
    new fetch_error(
      fetch_failure_kind.fetch_no_content,
      0,
      "no readable content extracted",
    ),
  ];
};

/*
PORT STATUS
source path: backend/web/extract.go
source lines: 267
confidence: high
status: phase_b_compile
*/
